StepLM

iulian453 opened the pull request:
Just finished the builtin function.

iulian453 committed:
initial commit of stepml

iulian453 committed:
first commit of steplm function

iulian453 committed:
second

iulian453 committed:
removed redundant variable intercept_status

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
```
Please put a "Modifications Copyright 2019 Graz University of Technology" similar to the lmCG builtin function on top.

mboehm7 commented on src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java:
```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2018 Graz University of Technology
```
Please change this to 2019 - the year when this file was added.

mboehm7 commented on src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.HashMap;
+
+
+public class BuiltinSTEPLmTest extends AutomatedTestBase{
```
Please change the indentation from spaces to tabs (we use for historic reasons tabs for code, and spaces for dml scripts).

mboehm7 commented on src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java:
```diff
@@ -0,0 +1,152 @@
+    @Test
+    public void testLmMatrixDenseSPlmDS() {
+        runLmTest(false, ExecType.SPARK, BuiltinLmTest.LinregType.DS);
+    }
```
Remove all unnecessary testcases - as in steplm you don't differentiate between DS, CG, or AUTO

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+#
+# HOW TO INVOKE THIS SCRIPT - EXAMPLE:
+# hadoop jar SystemML.jar -f StepLinearRegDS.dml -nvargs X=INPUT_DIR/X Y=INPUT_DIR/Y B=OUTPUT_DIR/betas
+#     O=OUTPUT_DIR/stats S=OUTPUT_DIR/selected icpt=2 thr=0.01 fmt=csv write_beta=TRUE
```
Remove this invocation example because it is rather misleading (we don't support MR anymore and this script is not invoked via command line anymore)

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+
+  #fmt = ifdef($fmt, "text");
+  #icpt = intercept_status;
+  thr = ifdef($thr, 0.001);
```
Remove all $-parameters from the script as they are command line arguments - either make them a function argument or remove entirely - I would prefer the former.

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+          column_best = i;
+          AIC_best = as.scalar(AICs[1, i]);
+        }
+        }
```
Here seems to be some indentation issue - please make a full pass over the script (which was not good in that regard) and make the formatting consistent. I'd recommend to enable whitespace characters in your IDE , then you'll see it at a first glance.

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+  } else {
+    stop("Currently only forward selection strategy is supported!");
+  }
+
```
remove unnecessary empty lines.

mboehm7 commented on src/test/scripts/functions/builtin/steplm.R:
```diff
@@ -0,0 +1,343 @@
+
+
+
+# Computes linear regression using a direct solver for (X^T X) beta = X^T y.
```
Please remove these spurious free lines and final comment.

mboehm7 commented on src/test/scripts/functions/builtin/steplm.dml:
```diff
@@ -0,0 +1,28 @@
+
+X = read($1)
+y = read($2)
+S = steplm(X = X, y = y, icpt = 1);
```
Please add a consumer of S (print(sum) or write) because otherwise it's subject to dead code elimination.

mboehm7 commented on scripts/builtin/steplm.dml:
```diff
@@ -0,0 +1,320 @@
+
+      write(Selected, $4);
+      write(beta_out, $3);
+   
```
Please make all write return arguments - this function should be (except for prints) side effect free.

mboehm7 reviewed (commented):
Thanks @iulian453 - that's a good start, it simply requires a couple of cleanups. Please let's me know once you addressed the comments and I'll have another look. Also, please make sure that the commit email address matches your email address of the github account; otherwise you contribution will not be corrected reflected.

iulian453 committed:
cleanups

iulian453 commented:
Hello Matthias,

I just addressed the comments, so please have a look and let me know if I missed something. 

P.D. The function I was using in Matlab for making the model is called "fitrsvm", and "resubloss" for predicting outputs.
P.D Just changed my the email in GitHub, so now all should be fine.

Regards,

Iulian 

> El 30 jul 2019, a las 23:29, Matthias Boehm <notifications@github.com> escribió:
> 
> @mboehm7 commented on this pull request.
> 
> Thanks @iulian453 <https://github.com/iulian453> - that's a good start, it simply requires a couple of cleanups. Please let's me know once you addressed the comments and I'll have another look. Also, please make sure that the commit email address matches your email address of the github account; otherwise you contribution will not be corrected reflected.
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308944807>:
> 
> > @@ -0,0 +1,320 @@
> +#-------------------------------------------------------------
> +#
> +# Licensed to the Apache Software Foundation (ASF) under one
> Please put a "Modifications Copyright 2019 Graz University of Technology" similar to the lmCG builtin function on top.
> 
> In src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308945783>:
> 
> > @@ -0,0 +1,152 @@
> +/*
> + * Copyright 2018 Graz University of Technology
> Please change this to 2019 - the year when this file was added.
> 
> In src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308946231>:
> 
> > + */
> +
> +package org.tugraz.sysds.test.functions.builtin;
> +
> +import org.junit.Test;
> +import org.tugraz.sysds.common.Types.ExecMode;
> +import org.tugraz.sysds.lops.LopProperties.ExecType;
> +import org.tugraz.sysds.runtime.matrix.data.MatrixValue.CellIndex;
> +import org.tugraz.sysds.test.AutomatedTestBase;
> +import org.tugraz.sysds.test.TestConfiguration;
> +import org.tugraz.sysds.test.TestUtils;
> +
> +import java.util.HashMap;
> +
> +
> +public class BuiltinSTEPLmTest extends AutomatedTestBase{
> Please change the indentation from spaces to tabs (we use for historic reasons tabs for code, and spaces for dml scripts).
> 
> In src/test/java/org/tugraz/sysds/test/functions/builtin/BuiltinSTEPLmTest.java <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308946661>:
> 
> > +    }
> +
> +    @Test
> +    public void testLmMatrixDenseCPlmDS() {
> +        runLmTest(false, ExecType.CP, BuiltinLmTest.LinregType.DS);
> +    }
> +
> +    @Test
> +    public void testLmMatrixSparseCPlmDS() {
> +        runLmTest(true, ExecType.CP, BuiltinLmTest.LinregType.DS);
> +    }
> +
> +    @Test
> +    public void testLmMatrixDenseSPlmDS() {
> +        runLmTest(false, ExecType.SPARK, BuiltinLmTest.LinregType.DS);
> +    }
> Remove all unnecessary testcases - as in steplm you don't differentiate between DS, CG, or AUTO
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308947164>:
> 
> > +# STDEV_RES_Y           Standard Deviation of the residual Y - pred(Y|X)
> +# DISPERSION            GLM-style dispersion, i.e. residual sum of squares / # deg. fr.
> +# R2                    R^2 of residual with bias included vs. total average
> +# ADJUSTED_R2           Adjusted R^2 of residual with bias included vs. total average
> +# R2_NOBIAS             R^2 of residual with bias subtracted vs. total average
> +# ADJUSTED_R2_NOBIAS    Adjusted R^2 of residual with bias subtracted vs. total average
> +# R2_VS_0               * R^2 of residual with bias included vs. zero constant
> +# ADJUSTED_R2_VS_0      * Adjusted R^2 of residual with bias included vs. zero constant
> +# -------------------------------------------------------------------------------------
> +# * The last two statistics are only printed if there is no intercept (icpt=0)
> +# If the best AIC is achieved without any features the matrix of selected features contains 0.  
> +# Moreover, in this case no further statistics will be produced  
> +#
> +# HOW TO INVOKE THIS SCRIPT - EXAMPLE:
> +# hadoop jar SystemML.jar -f StepLinearRegDS.dml -nvargs X=INPUT_DIR/X Y=INPUT_DIR/Y B=OUTPUT_DIR/betas
> +#     O=OUTPUT_DIR/stats S=OUTPUT_DIR/selected icpt=2 thr=0.01 fmt=csv write_beta=TRUE
> Remove this invocation example because it is rather misleading (we don't support MR anymore and this script is not invoked via command line anymore)
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308947588>:
> 
> > +# HOW TO INVOKE THIS SCRIPT - EXAMPLE:
> +# hadoop jar SystemML.jar -f StepLinearRegDS.dml -nvargs X=INPUT_DIR/X Y=INPUT_DIR/Y B=OUTPUT_DIR/betas
> +#     O=OUTPUT_DIR/stats S=OUTPUT_DIR/selected icpt=2 thr=0.01 fmt=csv write_beta=TRUE
> +
> +
> +m_steplm = function(Matrix[Double] X, Matrix[Double] y, Integer icpt = 0, Double reg = 1e-7, Double tol = 1e-7, Integer maxi = 0, Boolean verbose = TRUE)
> +  return() {
> +
> +
> +  # currently only the forward selection strategy in supported: start from one feature and iteratively add 
> +  # features until AIC improves
> +  dir = "forward";
> +
> +  #fmt = ifdef($fmt, "text");
> +  #icpt = intercept_status;
> +  thr = ifdef($thr, 0.001);
> Remove all $-parameters from the script as they are command line arguments - either make them a function argument or remove entirely - I would prefer the former.
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308948133>:
> 
> > +          tmp[1, 1] = i;
> +          columns_fixed_ordered_2 = append(columns_fixed_ordered, tmp);
> +      
> +          [AIC_2, beta_out_i] = linear_regression(X, y, icpt);
> +
> +          beta_out_all_2[1:nrow(beta_out_i), (i - 1) * y_ncol + 1:i * y_ncol] = beta_out_i[, 1:1];
> +
> +          AICs[1, i] = AIC_2;
> +
> +          AIC_cur = as.scalar(AICs[1, i]);
> +        if ((AIC_cur < AIC_best) & ((AIC_best - AIC_cur) > abs(thr * AIC_best)) &
> +            (as.scalar(columns_fixed[1, i]) == 0)) {
> +          column_best = i;
> +          AIC_best = as.scalar(AICs[1, i]);
> +        }
> +        }
> Here seems to be some indentation issue - please make a full pass over the script (which was not good in that regard) and make the formatting consistent. I'd recommend to enable whitespace characters in your IDE , then you'll see it at a first glance.
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308948240>:
> 
> > +    if (icpt != 0) {
> +      Selected = cbind(Selected, matrix(boa_ncol, rows = 1, cols = 1))
> +    }
> +
> +    beta_out = reorder_matrix(boa_ncol, beta_out, Selected);
> +
> +    print(toString(beta_out) + " mi beta out");
> +    print("Selected " + toString(Selected));
> +    write(Selected, $4);
> +    write(beta_out, $3);
> + 
> +
> +  } else {
> +    stop("Currently only forward selection strategy is supported!");
> +  }
> +
> remove unnecessary empty lines.
> 
> In src/test/scripts/functions/builtin/steplm.R <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308949967>:
> 
> > +}
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +
> +# Computes linear regression using a direct solver for (X^T X) beta = X^T y.
> Please remove these spurious free lines and final comment.
> 
> In src/test/scripts/functions/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308950322>:
> 
> > +# you may not use this file except in compliance with the License.
> +# You may obtain a copy of the License at
> +#
> +#   http://www.apache.org/licenses/LICENSE-2.0
> +#
> +# Unless required by applicable law or agreed to in writing, software
> +# distributed under the License is distributed on an "AS IS" BASIS,
> +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +# See the License for the specific language governing permissions and
> +# limitations under the License.
> +#
> +#-------------------------------------------------------------
> +
> +X = read($1)
> +y = read($2)
> +S = steplm(X = X, y = y, icpt = 1);
> Please add a consumer of S (print(sum) or write) because otherwise it's subject to dead code elimination.
> 
> In scripts/builtin/steplm.dml <https://github.com/tugraz-isds/systemds/pull/19#discussion_r308950609>:
> 
> > +        B = matrix(beta, rows = m_orig, cols = 1);
> +      } else {
> +        B_tmp = matrix(0, rows = m_orig + 1, cols = 1);
> +        B_tmp[m_orig + 1,] = beta;
> +        B = B_tmp;
> +      }
> +
> +      beta_out = B;
> +
> +      print("estoy en el stop");
> +      print(toString(beta_out) + " mi beta out");
> +      print("Selected " + toString(Selected));
> +
> +      write(Selected, $4);
> +      write(beta_out, $3);
> +   
> Please make all write return arguments - this function should be (except for prints) side effect free.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/tugraz-isds/systemds/pull/19?email_source=notifications&email_token=AHW7YFUKMDGK45TZD64VNBLQCCXCLA5CNFSM4IH4GHN2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCABZQKI#pullrequestreview-268671017>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHW7YFTK6FM54TEGQTNOIGTQCCXCLANCNFSM4IH4GHNQ>.
>

mboehm7 commented:
LGTM - that's awesome @iulian453, thanks so much for the patch. Also just for the future and in case my previous comment was misunderstood. I did not ask to remove all free lines but just the spurious ones (e.g., between closing braces).
